man/uncat.1
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "uncat" "1" 
.SH NAME
uncat \- continuously reads stdin, splits up the data either by timeout
or number of bytes read to process it
.SH SYNOPSIS
\f[B]uncat\f[R] [\-vo] [\-t \f[I]seconds\f[R]] [\-s \f[I]size\f[R]]
\f[I]prog\f[R]
.SH DESCRIPTION
\f[B]uncat\f[R] continuously reads stdin and repeatedly processes the
data either after \f[I]seconds\f[R] timeout or maximal number of bytes
\f[I]size\f[R] read.
To process the data, \f[B]uncat\f[R] starts \f[I]prog\f[R] and feeds the
data into \f[I]prog\f[R]\[aq]s standard input.
\f[I]prog\f[R]\[aq]s standard output is redirected to standard error.
\f[B]uncat\f[R] never writes to standard output.
.PP
\f[I]prog\f[R] consist of one or more arguments.
.PP
If \f[I]prog\f[R] exits non\-zero, \f[B]uncat\f[R] prints an error
message, discards the data and continues to read stdin.
.PP
Note that \f[B]uncat\f[R] is not crash proof.
.PP
\f[B]uncat\f[R] is used to run the \f[I]socklog\-notify\f[R] service.
.SH OPTIONS
.TP
\f[B]\-v\f[R]
verbose.
Print verbose messages to stderr.
.TP
\f[B]\-o\f[R]
once.
Exit on end of input.
Normally \f[B]uncat\f[R] continues waiting for data on standard input,
even on end of file.
.TP
\f[B]\-t\f[R] \f[I]seconds\f[R]
timeout.
Set the timeout to \f[I]seconds\f[R] seconds.
Default is 300.
.TP
\f[B]\-s\f[R] \f[I]size\f[R]
Set the maximal number of bytes to \f[I]size\f[R].
Default is 1024.
.SH SIGNALS
If \f[B]uncat\f[R] receives a TERM signal, it starts to process the data
in memory and exits as soon as possible.
.PP
\f[B]uncat\f[R] also does this on end of input if the \f[B]\-o\f[R]
option was given.
.SH EXIT CODES
\f[B]uncat\f[R] returns 0 after receiving a TERM signal.
\f[B]uncat\f[R] returns 111 on error.
.SH SEE ALSO
socklog(8), socklog\-conf(8), sv(8), runsv(8), svlogd(8), tryto(1)
.PP
https://smarden.org/socklog/
.PD 0
.P
.PD
https://smarden.org/runit/
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * doc/index.html: spelling.
  * doc/install.html: how to install as non-root user.
  * doc/examples.html, doc/notify.html: upper and lower case, format.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * socklog-check.c, socklog.c: check length of unix socket to match struct
    sockaddr.
  * pathexec_env.c, socklog-conf.c, socklog.c, tryto.c, uncat.c: whitespace.
  * socklog.c: fix compilation warning -Wpointer-sign.
  * socklog.c, man/socklog.8: add -T to drop timestamps from syslog messages
    (github/leahneukirchen).
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

socklog 2.1.2
Sun, 02 Nov 2025 11:41:11 +0100
  * alloc.c, alloc_re.c, buffer_0.c, byte_chr.c, byte_copy.c, byte_cr.c,
    byte_diff.c, byte_rchr.c, byte_zero.c, error.h, select.h2, wait.h,
    wait_nohang.c, wait_pid.c: sync daemontools library fork with runit
    source.
  * doc/*.html: !DOCTYPE.
  * socklog.c, tryto.c, uncat.c: use parameter name "unused" in sighandler
    functions.
  * hasflock.h1: add #include <unistd.h>.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-socklog-unix.c, pathexec_env.c, select.h1, select.h2,
    sig.c, sig.h, sig_catch.c, socklog-check.c, socklog-check.dist,
    socklog.c, tryto.c, tryto.dist, uncat.c, uncat.dist, wait.h: fix build
    with gcc15: in C23 f() means f(void), consequently signal handler
    functions now are f(int), alloc_free() and byte_zero() parameters need
    casts, buffer_unixwrite() removes "const" from parameter.
  * package/README: update copyright notice.
  * doc/configuration.html: replace links to daemontools faq with runit faq.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*, package/README: switch all smarden.org http links to https.
  * tryto.c: don't report failure if the final try of prog succeeds (thx
    Andras Korn).
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.

socklog 2.1.1
Fri, 11 Oct 2024 00:00:17 +0000
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * doc/index.html: remove reference to socklog 2.0.x.
  * doc/*.html: remove <small>$Id$</small>.
  * debian/: remove; obsolete.
  * pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c, sgetopt.c,
    sgetopt.h, socklog-check.c, socklog-conf.c, socklog.c, subgetopt.c,
    subgetopt.h, trycpp.c, tryflock.c, trypoll.c, trysgact.c, trysgprm.c,
    tryto.c, tryulong64.c, trywaitp.c, uncat.c, x86cpuid.c: incorporate
    lib changes from runit-2.2.0 and adapt socklog programs to properly
    build with modern toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * socklog.c: replace prot_gid() with setgroups() and prot_uid() with
    setuid().

socklog 2.1.0
Mon, 06 Mar 2006 12:59:33 +0000
  * socklog-conf, man/socklog-conf.8: create service directories
    /etc/sv/socklog-* instead of /etc/socklog/*; create ./check script
    for socklog-unix service.
  * doc/configuration.html, doc/install.html, doc/network.html,
    doc/notify.html: adapt to new path to service directories; utilize
    runit's sv program instead of runsvctrl.
  * man/socklog-check.8, man/socklog-conf.8, man/socklog.8, man/uncat.1:
    see also sv(8).

socklog 2.0.3
Sun, 26 Feb 2006 13:29:17 +0000
  * doc/index.html: minor.
  * socklog-check.c: change EDESTADDRREQ to ECONNREFUSED if defined and set
    by connect().
  * tryto: fix a race with child terminating immediatly (thx Alex Efros);
    catch SIGCHLD while waiting ktimeout seconds before sending KILL.
  * man/tryto.1: remove reference to timelimit(8) (thx Alex Efros).
  * socklog.check, socklog.dist: don't check usage output with -V as it
    differs on solaris (thx Asif Iqbal).

socklog 2.0.2
Sun, 10 Apr 2005 07:21:30 +0000
  * doc/readme.solaris.html: add -ldoor -lthread to src/conf-ld (thx Lars
    Kellogg-Stedman); update log/run script, mailing list address.
  * man/socklog.8: remove double entries in SEE ALSO.

socklog 2.0.1
Thu, 10 Feb 2005 10:05:55 +0000
  * Makefile: add linker options detected by trysocketlib.c to targets
    check-socklog-inet, check-socklog-inet (fixes test failure on solaris,
    thx Uffe Jakobsen).

socklog 2.0.0
Sun, 06 Feb 2005 19:57:04 +0000
  * socklog-check.check: tweak timing in program test.
  * tryto.c: new option -P: run prog in new session, and signal process
    group.
  * man/tryto.1: adapt; cleanup.
  * man/socklog-check.8: add hint about runit services using syslog
    facility; cleanup.
  * man/socklog-conf.8, man/socklog.8, man/uncat.1: cleanup.
  * doc/index.html: new stable release.
  * doc/network.html: typo.

socklog 1.5.0
Sat, 26 Jun 2004 11:28:33 +0000
  * socklog-check, man/socklog-check.8: new; check for the availability of
    a socklog service.
  * socklog-check.check, socklog-check.dist: new; program test.
  * byte_zero.c: new.
  * socklog.c, man/socklog.8: new option -U: don't override umask when
    creating domain socket.
  * package/check, package/compile, package/install-man, package/upgrade:
    minor cleanup.
  * Makefile, man/socklog-conf.8: adapt.

socklog 1.4.2
Tue, 23 Mar 2004 18:16:17 +0000
  * doc/index.html, doc/install.html: minor.
  * socklog.c: fix sigterm handling in ucspi mode.
  * uncat.check: tweak timing in program test.
  * Makefile: don't run target check by default.
  * package/check: new; run checks on programs.
  * package/install: run package/check.

socklog 1.4.1
Sat, 28 Feb 2004 17:47:13 +0000
  * socklog.c: handle sigterm on other platforms as accurately as already
    done on solaris.
  * tryto.c, uncat.c: minor cleanup.

src/uncat.c
#include <unistd.h>
#include "strerr.h"
#include "error.h"
#include "sgetopt.h"
#include "scan.h"
#include "stralloc.h"
#include "buffer.h"
#include "pathexec.h"
#include "fd.h"
#include "wait.h"
#include "taia.h"
#include "iopause.h"
#include "ndelay.h"
#include "sig.h"

/* defaults */
#define TIMEOUT 300
#define SIZEMAX 1024

#define USAGE " [-vo] [-t timeout] [-s size] prog"
#define WARNING "uncat: warning: "
#define FATAL "uncat: fatal: "

const char *progname;
int exitasap =0;

void exit_asap(int unused) {
  exitasap =1;
}

void usage () {
  strerr_die4x(1, "usage: ", progname, USAGE, "\n");
}

int main (int argc, char * const *argv, char * const *envp) {
  int opt;
  unsigned long timeout =TIMEOUT;
  unsigned long sizemax =SIZEMAX;
  int verbose =0;
  int once =0;
  static stralloc sa;
  int eof =0;

  progname =*argv;

  sig_block(sig_term);
  sig_catch(sig_term, exit_asap);

  while ((opt =getopt(argc, argv, "t:s:voV")) != opteof) {
    switch(opt) {
    case 'V':
      strerr_warn1("$Id$\n", 0);
    case '?':
      usage();
    case 't':
      scan_ulong(optarg, &timeout);
      if (timeout <= 0) timeout =TIMEOUT;
      break;
    case 's':
      scan_ulong(optarg, &sizemax);
      if (sizemax <= 0) sizemax =SIZEMAX;
      break;
    case 'v':
      verbose =1;
      break;
    case 'o':
      once =1;
      break;
    }
  }
  argv +=optind;
  if (!*argv) usage();
  
  if (verbose) strerr_warn2(WARNING, "starting.", 0);

  ndelay_on(0);

  for (;;) {
    struct taia now, deadline;
    int cpipe[2];
    int pid;
    int wstat;

    stralloc_copys(&sa, "");

    /* set timeout */
    taia_now(&now);
    taia_uint(&deadline, timeout);
    taia_add(&deadline, &now, &deadline);

    /* read fd 0, stop at maxsize bytes or timeout */
    for (;;) {
      int r;
      char *s;
      iopause_fd iofd;

      taia_now(&now);
      if (taia_less(&deadline, &now)) {
        if (verbose && sa.len) strerr_warn2(WARNING, "timeout reached.", 0);
        break;
      }
      iofd.fd =0;
      iofd.events =IOPAUSE_READ;

      sig_unblock(sig_term);
      iopause(&iofd, 1, &deadline, &now);
      sig_block(sig_term);
      
      if (exitasap) {
        if (verbose) strerr_warn2(WARNING, "got sigterm.", 0);
        break;
      }
      
      r =buffer_feed(buffer_0);
      if (r < 0) {
        if (errno == error_again) continue;
        strerr_die2sys(111, FATAL, "unable to read fd 0: ");
      }
      if (r == 0) {
        if (verbose) strerr_warn2(WARNING, "end of input.", 0);
        if (! once) continue;
        eof++;
        break;
      }
      if (r >= sizemax) r =sizemax;
      if ((sa.len +r) > sizemax) {
        if (verbose) strerr_warn2(WARNING, "max size reached.", 0);
        break;
      }
      s =buffer_peek(buffer_0);
      if (! stralloc_catb(&sa, s, r)) {
        strerr_die2sys(111, FATAL, "out of memory: ");
      }
      buffer_seek(buffer_0, r);
    }
    if (sa.len) {
      /* run prog to process sa.s */
      if (pipe(cpipe) == -1) {
        strerr_die2sys(111, FATAL, "unable to create pipe for child: ");
      }
      while ((pid =fork()) == -1) {
        strerr_warn4(WARNING, "unable to fork for \"", *argv, "\" pausing: ",
                     &strerr_sys);
        sleep(5);
      }
      if (!pid) {
        /* child */

        sig_uncatch(sig_term);
        sig_unblock(sig_term);

        close(cpipe[1]);
        fd_move(0, cpipe[0]);
        fd_copy(1, 2);

        if (verbose) strerr_warn2(WARNING, "starting child.", 0);
        pathexec_run(*argv, argv, envp);
        strerr_die2sys(111, FATAL, "unable to start child: ");
      }
      
      close(cpipe[0]);
      if (write(cpipe[1], sa.s, sa.len) < sa.len) {
        strerr_warn2(WARNING, "unable to write to child: ", &strerr_sys);
      }
      close(cpipe[1]);
      
      if (wait_pid(&wstat, pid) != pid) {
        strerr_die2sys(111, FATAL, "wait_pid: ");
      }
      if (wait_crashed(wstat)) {
        strerr_warn2(WARNING, "child crashed.", 0);
      } else {
        if (verbose) strerr_warn2(WARNING, "child exited.", 0);
      }
    }

    if (exitasap || eof) break;
  }

  if (verbose) strerr_warn2(WARNING, "exit.", 0);
  _exit(0);
}
